### PR TITLE
Defining sidebar ability

### PR DESCRIPTION
**Pull request by allenwq**
Filter sidebar items by user's ability

### CODE REVIEW

File: app/controllers/components/course/groups_component.rb [hunk 1]
```diff
@@ -12,4 +14,14 @@ def sidebar_items
...
   end
+
+  private
+
+  def can_manage_group?
+    can?(:manage, Course::Group.new(course: current_course)) || manageable_groups.any?
+  end
+
+  def manageable_groups
+    current_course.groups.select { |group| can?(:manage, group) }
```
Comment (lowjoel) [inline app/controllers/components/course/groups_component.rb]: accessible_by?

File: spec/features/course/sidebar_spec.rb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        expect(page).not_to have_selector('li', text: 'course.groups.sidebar_title')
+        expect(page).not_to have_selector('li', text: 'layouts.course_admin.title')
+      end
+    end
+
+    context 'As a Group Manager' do
+      let!(:group) { create(:course_group, course: course) }
+      let(:user) { create(:course_group_manager, course: course, course_group: group).user }
+
+      scenario 'I can view group sidebar item' do
```
Comment (lowjoel) [inline spec/features/course/sidebar_spec.rb]: the Groups Sidebar item

File: app/controllers/components/course/groups_component.rb (hunk 1 above)
Comment (allenwq) [inline app/controllers/components/course/groups_component.rb]: `accessible_by` is for viewing ?  but should also be fine in this case.

Comment (lowjoel) [inline app/controllers/components/course/groups_component.rb]: it should be the same right? accessible_by is filtering a set of candidate rows to one that matches the condition?

Comment (allenwq) [inline app/controllers/components/course/groups_component.rb]: Oh, I find `accessible_by` has a optional param: `def accessible_by(ability, action = :index)`, Thanks a lot !

File: app/controllers/components/course/groups_component.rb [hunk 3]
```diff
@@ -12,4 +14,14 @@ def sidebar_items
...
   end
+
+  private
+
+  def can_manage_group?
+    can?(:manage, Course::Group.new(course: current_course)) || manageable_groups.any?
+  end
+
+  def manageable_groups
+    current_course.groups.accessible_by(current_ability, :manage)
```
Comment (allenwq) [inline app/controllers/components/course/groups_component.rb]: @lowjoel Thanks a lot, `accessible_by(current_ability, :manage)` works

**Comment by lowjoel** [comment]
@allenwq I'm wondering if it's wise to lump all the sidebar specs together. The components framework are supposed to make them independent...

**Comment by allenwq** [comment]
@lowjoel  Actually initially I put a sidebar item spec in each of the feature spec, but then I thought they are slow (because for each component need few types of roles to test) and deleted all of them..

**Comment by allenwq** [comment]
But anyway I can put them back, yah I agree they are coupled now.

**Comment by lowjoel** [comment]
What you could do is the first feature spec for each feature will start from course home instead?

But I think if we test the sidebar individually as a component is good enough.

**Comment by allenwq** [comment]
You mean let feature spec start from course home and then click the sidebar item ? 

And this is good enough ? :)

**Comment by lowjoel** [comment]
Yes, and testing each component individually is okay.

**Comment by allenwq** [comment]
@lowjoel  I've added a sidebar item spec for each feature. And if all users in the course can see the sidebar item, I only tested the student role.

**Comment by allenwq** [comment]
@lowjoel Thanks !

**Pull request merged**